demux/demux_mkv.c

```diff
@@ -1510,16 +1505,23 @@ static int demux_mkv_open_video(demuxer_t *demuxer, mkv_track_t *track)
     sh_v->disp_w = track->v_width;
     sh_v->disp_h = track->v_height;
 
-    sh_v->crop.x0 = track->v_crop_left_set ? track->v_crop_left : 0;
-    sh_v->crop.y0 = track->v_crop_top_set ? track->v_crop_top : 0;
-    sh_v->crop.x1 = track->v_width -
-                        (track->v_crop_right_set ? track->v_crop_right : 0);
-    sh_v->crop.y1 = track->v_height -
-                        (track->v_crop_bottom_set ? track->v_crop_bottom : 0);
+    struct mp_rect crop;
+    crop.x0 = track->v_crop_left;
+    crop.y0 = track->v_crop_top;
+    crop.x1 = track->v_width - track->v_crop_right;
+    crop.y1 = track->v_height - track->v_crop_bottom;
 
-    int dw = track->v_dwidth_set ? track->v_dwidth : mp_rect_w(sh_v->crop);
-    int dh = track->v_dheight_set ? track->v_dheight : mp_rect_h(sh_v->crop);
-    struct mp_image_params p = {.w = mp_rect_w(sh_v->crop), .h = mp_rect_h(sh_v->crop)};
+    // Keep the codec crop rect as 0s if we have no cropping since the
+    // file may have broken width/height tags.
+    if (track->v_crop_left || track->v_crop_top ||
+        track->v_crop_right || track->v_crop_bottom)
+    {
+        sh_v->crop = crop;
+    }
+
+    int dw = track->v_dwidth_set ? track->v_dwidth : mp_rect_w(crop);
+    int dh = track->v_dheight_set ? track->v_dheight : mp_rect_h(crop);
+    struct mp_image_params p = {.w = mp_rect_w(crop), .h = mp_rect_h(crop)};
```

Comment:
Need to guard `mp_rect_*(crop)` calls. Something like this should work.

```suggestion
    bool crop_set = track->v_crop_left || track->v_crop_top || track->v_crop_right || track->v_crop_bottom;
    if (crop_set)
        sh_v->crop = crop;

    int w = crop_set ? mp_rect_w(crop) : track->v_width;
    int h = crop_set ? mp_rect_h(crop) : track->v_height;
    int dw = track->v_dwidth_set ? track->v_dwidth : w;
    int dh = track->v_dheight_set ? track->v_dheight : h;
    struct mp_image_params p = {.w = w, .h = h};
```

And while you are changing this please adjust `mp_image_to_av_frame` to support (0,0,0,0) crop, set the crop only if valid.

Reply:
>Need to guard mp_rect_*(crop) calls. 

Shouldn't it be the same? If there's no cropping then `mp_rect_w` is equal to `track->v_width` which is what we want. Ditto for the height case.

>And while you are changing this please adjust mp_image_to_av_frame to support (0,0,0,0) crop

I don't think this is needed? I'll admit I didn't 100% verify all of the code here, but I don't believe image param crop in `mp_image_to_av_frame` is the same as what's set in here. At least that's what the `video-params/crop-*` properties suggest. I assume that the calculation in `player/video.c` handles this already and makes the params the actual (0, 0, w, h) value which is what `mp_image_to_av_frame` and anything else that uses the `mp_image` gets.

Reply:
> Shouldn't it be the same? If there's no cropping then mp_rect_w is equal to track->v_width which is what we want. 

Yes, I'm not thinking late at night. I was focused that PAR will be wrong, but it will be wrong anyway, because actually the v_width is wrong, not the crop. I guess we have option to ignore that.

> I don't believe image param crop in mp_image_to_av_frame is the same as what's set in here.

True, I see I added special case for (0,0,0,0) crop and it does the job. Although I noticed one other issue, will fix tomorrow. And take a look again at your change, but looks good actually.